proparse/src/main/java/org/prorefactor/proparse/antlr4/TokenList.java

```diff
@@ -0,0 +1,191 @@
+
+  private final TokenSource tokenStream;
+  private final List<ProToken> list = new ArrayList<>();
+
+  private int currentPosition = 0;
+  private ProToken currentToken;
+  private boolean initialized = false;
+
+  TokenList(TokenSource input) {
+    this.tokenStream = input;
+  }
+
+  private void build()  {
+    LOGGER.trace("Entering TokenList#build()");
+    int zz = 0;
+    for (;;) {
+      ProToken nextToken = (ProToken) tokenStream.nextToken();
+      nextToken.setTokenIndex(zz++);
+      list.add(nextToken);
+      if (nextToken.getType() == PreprocessorParser.OBJCOLON)
+        reviewObjcolon();
+      if (nextToken.getType() == PreprocessorParser.EOF)
+        break;
+    }
+    LOGGER.trace("Exiting TokenList#build() - {} tokens", list.size());
+  }
+
+  /**
+   * Review the token list at an OBJCOLON token.
+   * 
+   * This is the reason this class was created in the first place. If we have an OBJCOLON, what comes before it has to
+   * be one of a few things:
+   * <ul>
+   * <li>a system handle,
+   * <li>a handle expression,
+   * <li>an Object reference expression, or
+   * <li>a type name (class name) for a static member reference
+   * </ul>
+   * <p>
+   * A type name can be pretty much anything, even a reserved keyword. It can also be a qualified class name, such as
+   * com.joanju.Foo.
+   * <p>
+   * This method attempts to resolve the following problem: Because of static class member references, a class name can
+   * be the first token in an expression. Class names can be composed of reserved keywords. This means that a reserved
+   * keyword could be the first piece of an expression, and this completely breaks the lookahead in the ANTLR generated
+   * parser. So, here, we look for OBJCOLON, and make sure that what comes before it is a system handle, an ID, or a
+   * non-reserved keyword.
+   * <p>
+   * A NAMEDOT token is a '.' followed by anything other than whitespace. If the OBJCOLON is proceeded by a NAMEDOT pair
+   * (NAMEDOT followed by anything), then we convert all of the NAMEDOT pairs to NAMEDOT-ID pairs. Otherwise, if the
+   * OBJCOLON is proceeded by any reserved keyword other than a systemhandlename, then we change that token's type to
+   * ID.
+   */
+  private void reviewObjcolon() {
+    int colonIndex = list.size() - 1;
+    int lastIndex = colonIndex - 1;
+
+    // Getting type of the token just before colon (excluding comments and whitespaces)
+    int ttype = list.get(lastIndex).getType();
+    while (ttype == PreprocessorParser.WS || ttype == PreprocessorParser.COMMENT) {
+      ttype = list.get(--lastIndex).getType();
+    }
+
+    // Look for NAMEDOT pairs.
+    // Actually, it's not that easy. Something like:
+    // newsyntax.101b.deep.FindMe
+    // is perfectly valid, and because of the digit following the '.',
+    // one part of that name gets picked up as a token with text ".101b".
+    int index = lastIndex;
+    boolean foundNamedot = false;
+    for (;;) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS135)